### PR TITLE
Add source table data into lineage header

### PR DESCRIPTION
**Pull request by kriscon-db**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/kriscon-db/mlflow/pull/10796?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/10796/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 10796
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?

- Altered delta_dataset_source.py to have the DeltaDatasetSource lookup the UC id and table name for any delta table source that resides within UC.  
- Added functionality to pass the newly captured UC id and table name into the lineage header that is passed to UCMR
- Updated protobuf definitions
- Added unit tests

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests

Screenshot 2024-01-17 at 8 54 14 PM

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [x] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/data/delta_dataset_source.py
Comment (smurching) [inline mlflow/data/delta_dataset_source.py:130]: Probably shouldn't be a warning since users can read delta tables in open source

File: mlflow/store/_unity_catalog/registry/rest_store.py
Comment (smurching) [inline mlflow/store/_unity_catalog/registry/rest_store.py:458]: Could these be accessed as properties prior to converting to `dict`? So we don't need to hardcode the strings here

File: mlflow/store/_unity_catalog/registry/rest_store.py
Comment (smurching) [inline mlflow/store/_unity_catalog/registry/rest_store.py:551]: Out of curiosity is there any limit on how long this list can be (how many MLflow datasets in a run)? Just so we don't pass a huge header in the request if there are too many tables. We could truncate if there's no limit

File: mlflow/store/_unity_catalog/registry/rest_store.py
Comment (prithvikannan) [reply mlflow/store/_unity_catalog/registry/rest_store.py:458]: @smurching we don't currently expose an attribute for this. the approach you described is hard to implement, since the returned object from `mlflow.data.get_source` is a `DatasetSource`, not a `DeltaDatasetSource`. i think this approach is okay.

File: mlflow/store/_unity_catalog/registry/rest_store.py
Comment (prithvikannan) [reply mlflow/store/_unity_catalog/registry/rest_store.py:551]: No limit on datasets per run today. Not sure how this backend service works, but I presume you have some server-side protection against large requests, which might be good enough here.

File: mlflow/store/_unity_catalog/registry/rest_store.py
Comment (prithvikannan) [inline mlflow/store/_unity_catalog/registry/rest_store.py:569]: nit but should this be a list like `lineages=[Lineage(...)]`?

File: mlflow/store/_unity_catalog/registry/rest_store.py
Comment (prithvikannan) [inline mlflow/store/_unity_catalog/registry/rest_store.py:566]: any reason why the `table_entity` is constructed here rather than in  `_get_table_details()`? that way we can avoid packing/unpacking this array of tuples.

File: mlflow/data/delta_dataset_source.py
Comment (prithvikannan) [inline mlflow/data/delta_dataset_source.py:130]: not familiar with the API request at hand, but do we want to add more descriptive error handling (ie. authentication error, invalid request, etc.)? maybe this is already taken care of with `rest_utils`

File: mlflow/data/delta_dataset_source.py
Comment (prithvikannan) [inline mlflow/data/delta_dataset_source.py:146]: if this lookup fails due to say 500 error, do we want this dataset logging to terminate? logged datasets are immutable, so this would result in an "incorrect" record logged.

**Review by prithvikannan — COMMENTED** [review]
PR overall looks good. left a few small comments, but the approach makes sense to me.

File: mlflow/store/_unity_catalog/registry/rest_store.py
Comment (kriscon-db) [reply mlflow/store/_unity_catalog/registry/rest_store.py:551]: I'll limit it to 10 for now.  We can revisit in the final CR.

File: mlflow/data/delta_dataset_source.py
Comment (kriscon-db) [reply mlflow/data/delta_dataset_source.py:146]: My intuition is to not fail logging if the id lookup fails.

File: mlflow/data/delta_dataset_source.py
Comment (kriscon-db) [reply mlflow/data/delta_dataset_source.py:130]: Sid suggests to get rid of this logging altogether, in an earlier comment.

**Comment by kriscon-db** [comment]
@mlflow-automation autoformat

File: mlflow/data/delta_dataset_source.py
Comment (smurching) [inline mlflow/data/delta_dataset_source.py:29]: Can we reuse code from elsewhere in MLflow for this? e.g. https://github.com/mlflow/mlflow/blob/ae27dedccd7d3b60ce79859c7755196d0e282083/mlflow/store/_unity_catalog/registry/utils.py#L29 instead of duplicating it

File: mlflow/data/delta_dataset_source.py
Comment (smurching) [inline mlflow/data/delta_dataset_source.py:135]: Can we reuse/share code with UC model name resolution for htis?

File: mlflow/store/_unity_catalog/registry/rest_store.py
Comment (smurching) [reply mlflow/store/_unity_catalog/registry/rest_store.py:459]: > the approach you described is hard to implement, since the returned object from mlflow.data.get_source is a DatasetSource, not a DeltaDatasetSource. i think this approach is okay.

Not sure it's that hard to implement, we could check the type of the object + access the property? I still prefer that to using the string field name here

File: tests/data/test_delta_dataset_source.py
Comment (smurching) [inline tests/data/test_delta_dataset_source.py:126]: Could we mock the `GetTable` API response instead of mocking the full method to get table name? We do want to test the new logic introduced here

File: tests/data/test_delta_dataset_source.py
Comment (smurching) [inline tests/data/test_delta_dataset_source.py:131]: Similarly can we mock something other than the full method here? Maybe we just need to mock `is_in_databricks_runtime` & force that to `True`

Comment (smurching) [reply tests/data/test_delta_dataset_source.py:131]: Or maybe the problem is we can't properly run this logic since open source delta doesn't support three-level table names?

File: tests/data/test_delta_dataset_source.py
Comment (smurching) [inline tests/data/test_delta_dataset_source.py:180]: It doesn't seem like we're testing that ID retrieval throws here

File: mlflow/data/delta_dataset_source.py
Comment (kriscon-db) [reply mlflow/data/delta_dataset_source.py:29]: I'll need to pull all of that out into a common util then as I don't think we should be importing from store/_unity_catalog/... here.

File: mlflow/data/delta_dataset_source.py
Comment (kriscon-db) [reply mlflow/data/delta_dataset_source.py:135]: Same as above.

File: tests/data/test_delta_dataset_source.py
Comment (kriscon-db) [reply tests/data/test_delta_dataset_source.py:130]: Let me see what I can do.  It seems like a lot of work to validate a straightforward check.  Additionally, you would still need to mock the table name resolve method, so what exactly are you testing here?  Set membership of the elements in DATABRICKS_LOCAL_METASTORE_NAMES and DATABRICKS_SAMPLES_CATALOG_NAME?

File: tests/data/test_delta_dataset_source.py
Comment (kriscon-db) [reply tests/data/test_delta_dataset_source.py:179]: Yeah I'm realizing that with the broken test result.  This was the last thing I had done last night so wasn't clearly thinking this through.  Will fix.

File: mlflow/store/_unity_catalog/registry/rest_store.py
Comment (kriscon-db) [reply mlflow/store/_unity_catalog/registry/rest_store.py:459]: FWIW, it looks like importing the DeltaDatasetSource here creates a circular import (tried doing the instanceAs test).

Comment (kriscon-db) [reply mlflow/store/_unity_catalog/registry/rest_store.py:459]: I can add these as constants into some util and reference them that way if you prefer that.

Comment (smurching) [reply mlflow/store/_unity_catalog/registry/rest_store.py:459]: Oh weird, I would prefer to resolve the circular import if it's not too hard - why is `DeltaDatasetSource` importing this file?

File: mlflow/store/_unity_catalog/registry/rest_store.py
Comment (smurching) [inline mlflow/store/_unity_catalog/registry/rest_store.py:575]:
```suggestion
```

File: tests/data/test_delta_dataset_source.py
Comment (smurching) [reply tests/data/test_delta_dataset_source.py:126]: Hm, was it not possible to address this?

File: tests/data/test_delta_dataset_source.py
Comment (smurching) [reply tests/data/test_delta_dataset_source.py:174]: (we can use `side_effect` instead of `return_value` in `mock.patch` to make the method throw)

File: tests/data/test_delta_dataset_source.py
Comment (smurching) [inline tests/data/test_delta_dataset_source.py:198]:
```suggestion
        side_effect=call_endpoint_impl,
```
We can replace this with `side_effect`, passing a function (could either be a function that returns the value we want or throws) to test the throwing case

File: tests/data/test_delta_dataset_source.py
Comment (smurching) [reply tests/data/test_delta_dataset_source.py:126]: Specifically I'm hoping for a test that verifies `call_endpoint` is called with the expected args; see e.g. https://github.com/mlflow/mlflow/blob/6e5ec775d1351ec3d8e88da87a317bdecb9f342d/tests/store/_unity_catalog/model_registry/test_unity_catalog_rest_store.py#L520-L593 for a similar example

File: tests/data/test_delta_dataset_source.py
Comment (kriscon-db) [reply tests/data/test_delta_dataset_source.py:174]: Returning None will cause a throw when the table_id is referenced in the response.  I can add the `call_endpoint` itself throwing, but there is no practical difference.

Comment (smurching) [reply tests/data/test_delta_dataset_source.py:174]: > Returning None will cause a throw when the table_id is referenced in the response. I can add the call_endpoint itself throwing, but there is no practical difference.

Sounds good to make `call_endpoint` throw, it's better since it's less subject to changes in how we handle the API response & more-realistically simulates what callers would see in reality

Comment (smurching) [reply tests/data/test_delta_dataset_source.py:174]: Another way to put it: in a typed language, we wouldn't be able to return `None` from `call_endpoint` unless we made the return value an `Option` or returned `null`. Better to match the original method signature (throw or return a response) when mocking it in tests

File: mlflow/store/_unity_catalog/registry/rest_store.py
Comment (smurching) [inline mlflow/store/_unity_catalog/registry/rest_store.py:454]: Can we check upfront whether there are more inputs than the max supported & log a warning? And avoid making a `GetTable` per table?

**Review by smurching — COMMENTED** [review]
Thanks @kriscon-db for the updates! LG once my last comments are addressed

**Comment by kriscon-db** [comment]
@mlflow-automation autoformat

**Comment by kriscon-db** [comment]
@mlflow-automation autoformat

**Review by smurching — APPROVED** [review]
Can we add a unit test for the truncation behavior (create run with >10 input datasets, verify the header only contains 10 of them)? LGTM after that's done

**Comment by kriscon-db** [comment]
> Can we add a unit test for the truncation behavior (create run with >10 input datasets, verify the header only contains 10 of them)? LGTM after that's done

@smurching Should we still inspect all input sources and filter down to just delta dataset sources and *then* truncate before we do the proto conversion?  I'm realizing that the reason I had it this way was to make sure we grabbed all of the deltaset data sources.  I don't think it makes sense to just grab the first 10 dataset sources which may not be delta datasets, in which case we still have to eat a scan of the input sources array and would only really be saving the extra proto conversion on delta datasets > 10.

**Comment by smurching** [comment]
> Should we still inspect all input sources and filter down to just delta dataset sources and then truncate before we do the proto conversion? I'm realizing that the reason I had it this way was to make sure we grabbed all of the delta data sources. I don't think it makes sense to just grab the first 10 dataset sources which may not be delta datasets, in which case we still have to eat a scan of the input sources array and would only really be saving the extra proto conversion on delta datasets > 10.

Your proposal sounds good, I'm not too concerned about spending time scanning arrays or doing proto conversions, more about limiting the size of the request header that gets sent containing table info & the number of API requests to `GetTable`. But I'm realizing the actual `GetTable` API requests happen one at a time when you log a UC table dataset via `mlflow.log_input` (?), which seems fine, so lineage header emission logic/truncation won't have any effect on the number of API requests that gets sent. 

We can also do the proto conversion upfront (like before) and then truncate afterwards as long as we log a warning when truncation is happening & make sure the end header isn't too large (which the unit test can catch)

**Comment by kriscon-db** [comment]
Sorry for the re-review request @smurching but I ended up having to do some no great things in the test that I want you to eyeball before we land this.

File: mlflow/store/_unity_catalog/registry/rest_store.py
Comment (smurching) [inline mlflow/store/_unity_catalog/registry/rest_store.py:469]: Could we make this message something that tells the user what's happening in context as they're creating a model version, e.g. `f"Model version has str(num_input_sources)} upstream datasets, which exceeds the max of 10 upstream datasets for lineage tracking. Only the first 10 datasets will be propagated to Unity Catalog lineage"

File: mlflow/store/_unity_catalog/registry/rest_store.py
Comment (smurching) [inline mlflow/store/_unity_catalog/registry/rest_store.py:491]: Should `_truncate_input_sources` be called here?

Comment (kriscon-db) [reply mlflow/store/_unity_catalog/registry/rest_store.py:491]: I meant to delete the method altogether.  Since we did not care about a scan or the proto generation, then it was just easier to include all of it in this method and I was able to make the ugly test that checks the truncation work.  Will update to delete the unused method.

Comment (kriscon-db) [reply mlflow/store/_unity_catalog/registry/rest_store.py:491]: We immediately filter out non-delta sources, and then filter away delta sources w/o table ids, so we should only be left with valid data source by the time we're hydrating protos.

Comment (smurching) [reply mlflow/store/_unity_catalog/registry/rest_store.py:491]: SGTM, maybe then we can just apply the warning message update here

**Review by smurching — APPROVED** [review]

**Comment by kriscon-db** [comment]
@mlflow-automation autoformat

**Comment by kriscon-db** [comment]
@mlflow-automation autoformat

**Pull request merged**
